### PR TITLE
Avoid scaling-down already terminating machines

### LINKED ISSUE #55: Do not scale-down the machines which are already being deleted.
**What happened**: We observed on Azure, the machine deletion took much longer than expected. During this time the node-object was registered, and autoscaler decided to scale-down this machine. As part of the scale-down, Autoscaler reduced the `replica` of the MachineDeployment by 1, which can lead to the removal of the healthy machine. Removing this machine could be unintended.

Ideally, Autoscaler should not trigger the scale-down of the machine which is already being deleted.

### PR DESCRIPTION
**Pull request by hardikdr**
**What this PR does / why we need it**: With this PR, we avoid scaling down the machines which are already being delelted.

**Which issue(s) this PR fixes**:
Fixes https://github.com/gardener/autoscaler/issues/55

**Special notes for your reviewer**:

**Release note**:

```improvement operator
Autoscaler avoids scaling down the machines which are already being terminated.
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go [hunk 1]
```diff
@@ -301,7 +302,14 @@ func (m *McmManager) DeleteMachines(machines []*Ref) error {
 
 			mclone := mach.DeepCopy()
 
+			if isMachineTerminating(mclone) {
+				terminatingMachines = append(terminatingMachines, mclone)
+			}
 			if mclone.Annotations != nil {
+				if mclone.Annotations["machinepriority.machine.sapcloud.io"] == "1" {
```
Comment (prashanth26) [inline cluster-autoscaler/cloudprovider/mcm/mcm_manager.go]: Minor suggestion. Use a string constant here?

**Review by prashanth26 — APPROVED** [review]
/lgtm

File: cluster-autoscaler/cloudprovider/mcm/mcm_manager.go (hunk 1 above)
Comment (hardikdr) [reply cluster-autoscaler/cloudprovider/mcm/mcm_manager.go]: yes, done. Thanks.

**Comment by hardikdr** [comment]
/hold
As discussed, I'd like to rely on `DeletionTimestamp` and not phase `Terminating` to better safeguard us in cases where MCM is not running. I'll make the changes soon.

**Pull request merged**